{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Exploratory data analysis for Scicore- DataEngineering \n",
    "\n",
    "Inital data exploration of the 4 datasets and rlevant attributes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pydantic\n",
    "from typing import List, Tuple\n",
    "from datetime import datetime\n",
    "from pathlib import Path\n",
    "from io import StringIO\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Illumina data\n",
    "\n",
    "relevant columns are 1) sample number, 2) path to R1 file, 3) path to R2 file"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## description\n",
    "\n",
    "By looking at the illumina.tree file it is clear there are several different types of files:    \n",
    "1. R1/R2 \"fq.qz\" files which should be FASTQ genetic sequencing data\n",
    "2. R1/R2 \"trimming_report.txt\"  files which could be reports on the trimming operation to create the \"short reads\" genetic sequencing\n",
    "3. slurm log file from the trimSlurmReports, probably related to some preprocessing operation on the data. It seems there is no easy way to connect the slurmreports to the sequenecing files above\n",
    "\n",
    "I will create a dataframe for each file with the following columns:\n",
    "sample_id, sequence, length, read_type, filepath only for the FASTQ data, exluding trimming reports and slurm jobs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pydantic data model for illumina data\n",
    "\n",
    "class IlluminaDataModel(pydantic.BaseModel):\n",
    "    sample_id: str\n",
    "    sample_id_suffix: str\n",
    "    seq: str\n",
    "    sample_len: str\n",
    "    read_type: str\n",
    "    filename: str\n",
    "\n",
    "    @pydantic.validator('sample_id')\n",
    "    def clean_sample_id(cls,v):\n",
    "        \"\"\" cleans the sample id removing non standard characters\"\"\"\n",
    "        clean_id = v.split(\" \")[-1]\n",
    "        if not all([a.isnumeric() for a in clean_id.split(\"-\")]):\n",
    "            raise ValueError(f\"sample id is non standard: {clean_id}\")\n",
    "        else:\n",
    "            return clean_id.split(\"-\")[0]\n",
    "        \n",
    "    @pydantic.validator('sample_id_suffix')\n",
    "    def clean_sample_id_suffix(cls,v):\n",
    "        \"\"\" cleans the sample id removing non standard characters\"\"\"\n",
    "        clean_id = v.split(\" \")[-1]\n",
    "        if not all([a.isnumeric() for a in clean_id.split(\"-\")]):\n",
    "            raise ValueError(f\"sample id is non standard: {clean_id}\")\n",
    "        else:\n",
    "            return \"-\".join(clean_id.split(\"-\")[1:])\n",
    "        \n",
    "\n",
    "    @pydantic.validator('seq')\n",
    "    def check_sequence(cls,v):\n",
    "        \"\"\" checks that the sequence is in format letter:digit(s)\"\"\"\n",
    "        if (v[0].isnumeric()) or (not v[1:].isnumeric()):\n",
    "            raise ValueError( f\"sequence is non standard: {v}\")\n",
    "        else:\n",
    "            return v\n",
    "        \n",
    "    @pydantic.validator('sample_len')\n",
    "    def check_length(cls,v):\n",
    "        \"\"\" checks that the sequence is in format letter:digit(s)\"\"\"\n",
    "        if (v[0].isnumeric()) or (not v[1:-1].isnumeric()):\n",
    "            raise ValueError( f\"sequence is non standard: {v}\")\n",
    "        else:\n",
    "            return v\n",
    "        \n",
    "    @pydantic.validator('read_type')\n",
    "    def clean_read_type(cls,v):\n",
    "        \"\"\" checks that the sequence is in format letter:digit(s)\"\"\"\n",
    "        if v[0].isnumeric() or (not v[1:].isnumeric()):\n",
    "            raise ValueError( f\"sequence is non standard: {v}\")\n",
    "        else:\n",
    "            return int(v[1:])\n",
    "\n",
    "    \n",
    "    @pydantic.validator('filename')\n",
    "    def check_filename(cls, v):\n",
    "        \"\"\"check filename for correct extension ans beginning\"\"\"\n",
    "        clean_fname= v.split(\" \")[-1]\n",
    "        if (not clean_fname.endswith(\".gz\")) or (not clean_fname[0].isnumeric()):\n",
    "            raise ValueError(f\"filename is non standard: {clean_fname}\")\n",
    "        else:\n",
    "            return clean_fname\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total number of objects imported correctly: 336\n",
      "total number of errors: 0\n"
     ]
    }
   ],
   "source": [
    "def import_and_check_ill_data(\n",
    "    fp: str, dataModel: pydantic.BaseModel\n",
    ") -> Tuple[List[pydantic.BaseModel], List[dict]]:\n",
    "    \"\"\"opens a tree file and imports the data following given\n",
    "    pydantic data model\n",
    "\n",
    "    Returns:\n",
    "    list of objects as per data model, list of errors\n",
    "\n",
    "    \"\"\"\n",
    "    data = []\n",
    "    errors = []\n",
    "    # Open the .tree file and read its contents\n",
    "    with open(fp, \"r\") as f:\n",
    "        for line in f:\n",
    "            line = line.rstrip(\"\\n\")\n",
    "\n",
    "            if line.endswith(\"gz\"):\n",
    "                components = line.strip().split(\"_\")\n",
    "                try:\n",
    "                    obj = dataModel(\n",
    "                        sample_id=components[0],\n",
    "                        sample_id_suffix=components[0],\n",
    "                        seq=components[1],\n",
    "                        sample_len=components[2],\n",
    "                        read_type=components[3],\n",
    "                        filename=line,\n",
    "                    )\n",
    "                    data.append(obj)\n",
    "                except pydantic.ValidationError as e:\n",
    "                    errors.append({\"line\": line, \"error\": str(e)})\n",
    "\n",
    "    return data, errors\n",
    "\n",
    "\n",
    "# ----\n",
    "fp = \"../illumina_data.tree\"\n",
    "objs, errors = import_and_check_ill_data(fp, IlluminaDataModel)\n",
    "print(f\"total number of objects imported correctly: {len(objs)}\")\n",
    "print(f\"total number of errors: {len(errors)}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 336 entries, 0 to 335\n",
      "Data columns (total 6 columns):\n",
      " #   Column            Non-Null Count  Dtype \n",
      "---  ------            --------------  ----- \n",
      " 0   sample_id         336 non-null    object\n",
      " 1   sample_id_suffix  336 non-null    object\n",
      " 2   seq               336 non-null    object\n",
      " 3   sample_len        336 non-null    object\n",
      " 4   read_type         336 non-null    int64 \n",
      " 5   filename          336 non-null    object\n",
      "dtypes: int64(1), object(5)\n",
      "memory usage: 15.9+ KB\n"
     ]
    }
   ],
   "source": [
    "# load illumina_data as dataframe for exploration\n",
    "ill_df = pd.DataFrame([a.dict() for a in objs])\n",
    "ill_df.info()\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Nanopore data\n",
    "\n",
    "relevant columns are 1) sequencing date, 2) path to sequencing files (one per row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "# nanopore datamodel\n",
    "class NanoporeDataModel(pydantic.BaseModel):\n",
    "    seq_date: str\n",
    "    barcode_number: str\n",
    "    path_to_seq_file: str\n",
    "\n",
    "    @pydantic.validator(\"seq_date\")\n",
    "    def transform_date(cls, v):\n",
    "        \"\"\"transforms str to date\"\"\"\n",
    "        clean_date = datetime.strptime(v, \"%Y%m%d\")\n",
    "        return clean_date\n",
    "\n",
    "    @pydantic.validator(\"barcode_number\")\n",
    "    def transform_barcode(cls, v):\n",
    "        if not v.isnumeric():\n",
    "            raise ValueError(f\"barcode is not numeric: {v}\")\n",
    "        return int(v)\n",
    "\n",
    "    @pydantic.validator(\"path_to_seq_file\")\n",
    "    def check_path(cls, v):\n",
    "        \"\"\"checks that path is actually a consistent path using pathlib\"\"\"\n",
    "        my_path = Path(v)\n",
    "        return v\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total number of objects imported correctly: 194\n",
      "total number of errors: 20\n"
     ]
    }
   ],
   "source": [
    "def load_and_check_nanopore_data(\n",
    "    filepath: str, dataModel: pydantic.BaseModel\n",
    ") -> Tuple[List[pydantic.BaseModel], List[dict]]:\n",
    "    \"\"\"loads nanopore data from source file checking consistency using the given pydantic data model\"\"\"\n",
    "\n",
    "    data = []\n",
    "    errors = []\n",
    "    # Open the .tree file and read its contents\n",
    "    with open(filepath, \"r\") as f:\n",
    "        for line in f:\n",
    "            line = line.rstrip(\"\\n\")\n",
    "            if line.startswith(\"nanopore\"):\n",
    "                parent = line\n",
    "                seq_date = parent.split(\"/\")[1].split(\"_\")[0]\n",
    "            else:\n",
    "                components = line.split(\" \")[-1].split(\".\")\n",
    "                try:\n",
    "                    obj = dataModel(\n",
    "                        seq_date=seq_date,\n",
    "                        barcode_number=components[0][-2:],\n",
    "                        path_to_seq_file=f\"{parent}/{line.split(' ')[-1]}\",\n",
    "                    )\n",
    "                    data.append(obj)\n",
    "                except pydantic.ValidationError as e:\n",
    "                    errors.append({\"line\": line, \"date\": seq_date, \"error\": str(e)})\n",
    "    return data, errors\n",
    "\n",
    "\n",
    "# ---\n",
    "\n",
    "filepath = \"../nanopore_data.tree\"\n",
    "objs, errors = load_and_check_nanopore_data(filepath, NanoporeDataModel)\n",
    "print(f\"total number of objects imported correctly: {len(objs)}\")\n",
    "print(f\"total number of errors: {len(errors)}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 194 entries, 0 to 193\n",
      "Data columns (total 3 columns):\n",
      " #   Column            Non-Null Count  Dtype         \n",
      "---  ------            --------------  -----         \n",
      " 0   seq_date          194 non-null    datetime64[ns]\n",
      " 1   barcode_number    194 non-null    int64         \n",
      " 2   path_to_seq_file  194 non-null    object        \n",
      "dtypes: datetime64[ns](1), int64(1), object(1)\n",
      "memory usage: 4.7+ KB\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_date</th>\n",
       "      <th>barcode_number</th>\n",
       "      <th>path_to_seq_file</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2018-01-04</td>\n",
       "      <td>5</td>\n",
       "      <td>nanopore/20180104_Carbapenemases/porechop/BC05...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2018-01-04</td>\n",
       "      <td>6</td>\n",
       "      <td>nanopore/20180104_Carbapenemases/porechop/BC06...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2018-01-04</td>\n",
       "      <td>7</td>\n",
       "      <td>nanopore/20180104_Carbapenemases/porechop/BC07...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2018-01-04</td>\n",
       "      <td>8</td>\n",
       "      <td>nanopore/20180104_Carbapenemases/porechop/BC08...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2018-01-04</td>\n",
       "      <td>9</td>\n",
       "      <td>nanopore/20180104_Carbapenemases/porechop/BC09...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    seq_date  barcode_number  \\\n",
       "0 2018-01-04               5   \n",
       "1 2018-01-04               6   \n",
       "2 2018-01-04               7   \n",
       "3 2018-01-04               8   \n",
       "4 2018-01-04               9   \n",
       "\n",
       "                                    path_to_seq_file  \n",
       "0  nanopore/20180104_Carbapenemases/porechop/BC05...  \n",
       "1  nanopore/20180104_Carbapenemases/porechop/BC06...  \n",
       "2  nanopore/20180104_Carbapenemases/porechop/BC07...  \n",
       "3  nanopore/20180104_Carbapenemases/porechop/BC08...  \n",
       "4  nanopore/20180104_Carbapenemases/porechop/BC09...  "
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nano_df = pd.DataFrame([a.dict() for a in objs])\n",
    "nano_df.info()\n",
    "nano_df.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20180104 1 validation error for NanoporeDataModel\n",
      "barcode_number\n",
      "  barcode is not numeric: ne (type=value_error) \n",
      " └── none.fastq.gz\n",
      "20180108 1 validation error for NanoporeDataModel\n",
      "barcode_number\n",
      "  barcode is not numeric: ne (type=value_error) \n",
      " └── none.fastq.gz\n",
      "20180110 1 validation error for NanoporeDataModel\n",
      "barcode_number\n",
      "  barcode is not numeric: ne (type=value_error) \n",
      " └── none.fastq.gz\n",
      "20180118 1 validation error for NanoporeDataModel\n",
      "barcode_number\n",
      "  barcode is not numeric: ne (type=value_error) \n",
      " └── none.fastq.gz\n",
      "20180123 1 validation error for NanoporeDataModel\n",
      "barcode_number\n",
      "  barcode is not numeric: ne (type=value_error) \n",
      " └── none.fastq.gz\n",
      "20180126 1 validation error for NanoporeDataModel\n",
      "barcode_number\n",
      "  barcode is not numeric: ne (type=value_error) \n",
      " ├── none.fastq.gz\n",
      "20180126 1 validation error for NanoporeDataModel\n",
      "barcode_number\n",
      "  barcode is not numeric: es (type=value_error) \n",
      " └── samples.txt\n",
      "20180208 1 validation error for NanoporeDataModel\n",
      "barcode_number\n",
      "  barcode is not numeric: ne (type=value_error) \n",
      " └── none.fastq.gz\n",
      "20180212 1 validation error for NanoporeDataModel\n",
      "barcode_number\n",
      "  barcode is not numeric: ne (type=value_error) \n",
      " └── none.fastq.gz\n",
      "20180215 1 validation error for NanoporeDataModel\n",
      "barcode_number\n",
      "  barcode is not numeric: ne (type=value_error) \n",
      " └── none.fastq.gz\n",
      "20180222 1 validation error for NanoporeDataModel\n",
      "barcode_number\n",
      "  barcode is not numeric: ne (type=value_error) \n",
      " └── none.fastq.gz\n",
      "20180301 1 validation error for NanoporeDataModel\n",
      "barcode_number\n",
      "  barcode is not numeric: ne (type=value_error) \n",
      " └── none.fastq.gz\n",
      "20180305 1 validation error for NanoporeDataModel\n",
      "barcode_number\n",
      "  barcode is not numeric: ne (type=value_error) \n",
      " └── none.fastq.gz\n",
      "20180309 1 validation error for NanoporeDataModel\n",
      "barcode_number\n",
      "  barcode is not numeric: ne (type=value_error) \n",
      " └── none.fastq.gz\n",
      "20180320 1 validation error for NanoporeDataModel\n",
      "barcode_number\n",
      "  barcode is not numeric: ne (type=value_error) \n",
      " └── none.fastq.gz\n",
      "20180323 1 validation error for NanoporeDataModel\n",
      "barcode_number\n",
      "  barcode is not numeric: ne (type=value_error) \n",
      " └── none.fastq.gz\n",
      "20180323 1 validation error for NanoporeDataModel\n",
      "barcode_number\n",
      "  barcode is not numeric: ne (type=value_error) \n",
      " └── none.fastq.gz\n",
      "20180406 1 validation error for NanoporeDataModel\n",
      "barcode_number\n",
      "  barcode is not numeric: ne (type=value_error) \n",
      " └── none.fastq.gz\n",
      "20180420 1 validation error for NanoporeDataModel\n",
      "barcode_number\n",
      "  barcode is not numeric:  (type=value_error) \n",
      " \n",
      "20180420 1 validation error for NanoporeDataModel\n",
      "barcode_number\n",
      "  barcode is not numeric: es (type=value_error) \n",
      " rectories, 212 files\n"
     ]
    }
   ],
   "source": [
    "for e in errors:\n",
    "    print(e[\"date\"][-20:], e[\"error\"], \"\\n\", e[\"line\"][-20:])\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MIC data\n",
    "\n",
    "relevant columns are 1) sample number, 2) path to R1 file, 3) path to R2 file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strain</th>\n",
       "      <th>Isolate #</th>\n",
       "      <th>Species</th>\n",
       "      <th>Labor Number</th>\n",
       "      <th>Date</th>\n",
       "      <th>Box</th>\n",
       "      <th>Platz</th>\n",
       "      <th>DNA concentration</th>\n",
       "      <th>Date of isolation</th>\n",
       "      <th>Localisation</th>\n",
       "      <th>...</th>\n",
       "      <th>Norfloxacin</th>\n",
       "      <th>Ciprofloxacin</th>\n",
       "      <th>Levofloxacin</th>\n",
       "      <th>Colistin</th>\n",
       "      <th>Fosfomycin-Trometamol</th>\n",
       "      <th>Tigecyclin</th>\n",
       "      <th>Rifampicin</th>\n",
       "      <th>Minocyclin</th>\n",
       "      <th>Doxycyclin</th>\n",
       "      <th>comment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>carb001</td>\n",
       "      <td>38</td>\n",
       "      <td>Kleb</td>\n",
       "      <td>402738</td>\n",
       "      <td>2010</td>\n",
       "      <td>1.0</td>\n",
       "      <td>A2</td>\n",
       "      <td>27.4</td>\n",
       "      <td>11/15/2010</td>\n",
       "      <td>unknown (culture)</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Phenotypic evidence for carbapenemase; KPC, VI...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>carb002</td>\n",
       "      <td>48</td>\n",
       "      <td>Kleb</td>\n",
       "      <td>402859</td>\n",
       "      <td>2010</td>\n",
       "      <td>1.0</td>\n",
       "      <td>A6</td>\n",
       "      <td>18.2</td>\n",
       "      <td>12/1/2010</td>\n",
       "      <td>unknown (culture)</td>\n",
       "      <td>...</td>\n",
       "      <td>&gt;=16</td>\n",
       "      <td>&gt;=4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&gt;=256</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Phenotypic evidence for carbapenemase; KPC, VI...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>carb003</td>\n",
       "      <td>14</td>\n",
       "      <td>Esch</td>\n",
       "      <td>800329</td>\n",
       "      <td>2011</td>\n",
       "      <td>1.0</td>\n",
       "      <td>B3</td>\n",
       "      <td>57.6</td>\n",
       "      <td>1/26/2011</td>\n",
       "      <td>Swab, inguinal</td>\n",
       "      <td>...</td>\n",
       "      <td>&gt;=16</td>\n",
       "      <td>&gt;=4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.125</td>\n",
       "      <td>&lt;=16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>no genotypic PCR result available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>carb004</td>\n",
       "      <td>41</td>\n",
       "      <td>Kleb</td>\n",
       "      <td>400793</td>\n",
       "      <td>2011</td>\n",
       "      <td>1.0</td>\n",
       "      <td>B5</td>\n",
       "      <td>39.6</td>\n",
       "      <td>3/22/2011</td>\n",
       "      <td>unknown (culture)</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;=16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KPC in PCR detected</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>carb005</td>\n",
       "      <td>31</td>\n",
       "      <td>Kleb</td>\n",
       "      <td>704984</td>\n",
       "      <td>2011</td>\n",
       "      <td>1.0</td>\n",
       "      <td>B6</td>\n",
       "      <td>28.6</td>\n",
       "      <td>3/28/2011</td>\n",
       "      <td>Urine</td>\n",
       "      <td>...</td>\n",
       "      <td>&gt;=16</td>\n",
       "      <td>&gt;=4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5</td>\n",
       "      <td>&gt;=256</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Phenotypic ESBL and porinlost</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 45 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    strain  Isolate # Species Labor Number  Date  Box Platz  \\\n",
       "0  carb001         38    Kleb       402738  2010  1.0    A2   \n",
       "1  carb002         48    Kleb       402859  2010  1.0    A6   \n",
       "2  carb003         14    Esch       800329  2011  1.0    B3   \n",
       "3  carb004         41    Kleb       400793  2011  1.0    B5   \n",
       "4  carb005         31    Kleb       704984  2011  1.0    B6   \n",
       "\n",
       "   DNA concentration Date of isolation       Localisation  ...  Norfloxacin  \\\n",
       "0               27.4        11/15/2010  unknown (culture)  ...          NaN   \n",
       "1               18.2         12/1/2010  unknown (culture)  ...         >=16   \n",
       "2               57.6         1/26/2011     Swab, inguinal  ...         >=16   \n",
       "3               39.6         3/22/2011  unknown (culture)  ...          NaN   \n",
       "4               28.6         3/28/2011              Urine  ...         >=16   \n",
       "\n",
       "  Ciprofloxacin Levofloxacin Colistin Fosfomycin-Trometamol  Tigecyclin  \\\n",
       "0           NaN          NaN      NaN                   NaN         NaN   \n",
       "1           >=4          NaN      NaN                 >=256         NaN   \n",
       "2           >=4          NaN    0.125                  <=16         NaN   \n",
       "3           NaN          NaN      NaN                  <=16         NaN   \n",
       "4           >=4          NaN      0.5                 >=256         NaN   \n",
       "\n",
       "  Rifampicin Minocyclin Doxycyclin  \\\n",
       "0        NaN        NaN        NaN   \n",
       "1        NaN        NaN        NaN   \n",
       "2        NaN        NaN        NaN   \n",
       "3        NaN        NaN        NaN   \n",
       "4        NaN        NaN        NaN   \n",
       "\n",
       "                                             comment  \n",
       "0  Phenotypic evidence for carbapenemase; KPC, VI...  \n",
       "1  Phenotypic evidence for carbapenemase; KPC, VI...  \n",
       "2                  no genotypic PCR result available  \n",
       "3                                KPC in PCR detected  \n",
       "4                      Phenotypic ESBL and porinlost  \n",
       "\n",
       "[5 rows x 45 columns]"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fp = \"../MIC_data.csv\"\n",
    "\n",
    "mic_df = pd.read_csv(fp)\n",
    "\n",
    "mic_df.head()\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sample TABLE\n",
    "\n",
    "relevant columns are 1) sample number, 2) path to R1 file, 3) path to R2 file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Internal #</th>\n",
       "      <th>Isolate #</th>\n",
       "      <th>Species</th>\n",
       "      <th>Labor Number</th>\n",
       "      <th>Date</th>\n",
       "      <th>Box</th>\n",
       "      <th>Platz</th>\n",
       "      <th>DNA Konzentration</th>\n",
       "      <th>sequenced</th>\n",
       "      <th>mb</th>\n",
       "      <th>barcode</th>\n",
       "      <th>resequenced</th>\n",
       "      <th>rebarcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>38</td>\n",
       "      <td>Kleb</td>\n",
       "      <td>402738</td>\n",
       "      <td>2010</td>\n",
       "      <td>1.0</td>\n",
       "      <td>A2</td>\n",
       "      <td>27.4</td>\n",
       "      <td>20180104</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>48</td>\n",
       "      <td>Kleb</td>\n",
       "      <td>402859</td>\n",
       "      <td>2010</td>\n",
       "      <td>1.0</td>\n",
       "      <td>A6</td>\n",
       "      <td>18.2</td>\n",
       "      <td>20180123</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7</td>\n",
       "      <td>20180420.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>14</td>\n",
       "      <td>Esch</td>\n",
       "      <td>800329</td>\n",
       "      <td>2011</td>\n",
       "      <td>1.0</td>\n",
       "      <td>B3</td>\n",
       "      <td>57.6</td>\n",
       "      <td>20180104</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>41</td>\n",
       "      <td>Kleb</td>\n",
       "      <td>400793</td>\n",
       "      <td>2011</td>\n",
       "      <td>1.0</td>\n",
       "      <td>B5</td>\n",
       "      <td>39.6</td>\n",
       "      <td>20180104</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>31</td>\n",
       "      <td>Kleb</td>\n",
       "      <td>704984</td>\n",
       "      <td>2011</td>\n",
       "      <td>1.0</td>\n",
       "      <td>B6</td>\n",
       "      <td>28.6</td>\n",
       "      <td>20180104</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Internal #  Isolate # Species Labor Number  Date  Box Platz  \\\n",
       "0          1         38    Kleb       402738  2010  1.0    A2   \n",
       "1          2         48    Kleb       402859  2010  1.0    A6   \n",
       "2          3         14    Esch       800329  2011  1.0    B3   \n",
       "3          4         41    Kleb       400793  2011  1.0    B5   \n",
       "4          5         31    Kleb       704984  2011  1.0    B6   \n",
       "\n",
       "   DNA Konzentration sequenced  mb barcode  resequenced  rebarcode  \n",
       "0               27.4  20180104 NaN       5          NaN        NaN  \n",
       "1               18.2  20180123 NaN       7   20180420.0        2.0  \n",
       "2               57.6  20180104 NaN       6          NaN        NaN  \n",
       "3               39.6  20180104 NaN       7          NaN        NaN  \n",
       "4               28.6  20180104 NaN       8          NaN        NaN  "
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fp = \"../sample_table.csv\"\n",
    "\n",
    "sample_df = pd.read_csv(fp)\n",
    "\n",
    "sample_df.head()\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# exploring uniProt api\n",
    "\n",
    "Here we build a query for the restapi of uniprot, testing to get data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "\n",
    "url = \"https://rest.uniprot.org/uniprotkb/stream?compressed=false&format=fasta&query=%28%28taxonomy_id%3A573%29%20AND%20Imipenem%29\"\n",
    "all_fastas = requests.get(url).text\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>description</th>\n",
       "      <th>name</th>\n",
       "      <th>seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tr|A0A0A8IKC4|A0A0A8IKC4_KLEPN</td>\n",
       "      <td>tr|A0A0A8IKC4|A0A0A8IKC4_KLEPN Bifunctional NA...</td>\n",
       "      <td>tr|A0A0A8IKC4|A0A0A8IKC4_KLEPN</td>\n",
       "      <td>(M, M, D, H, T, M, T, K, N, P, D, S, I, P, Y, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sp|C7C422|BLAN1_KLEPN</td>\n",
       "      <td>sp|C7C422|BLAN1_KLEPN Metallo-beta-lactamase t...</td>\n",
       "      <td>sp|C7C422|BLAN1_KLEPN</td>\n",
       "      <td>(M, E, L, P, N, I, M, H, P, V, A, K, L, S, T, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>tr|Q6XEC0|Q6XEC0_KLEPN</td>\n",
       "      <td>tr|Q6XEC0|Q6XEC0_KLEPN Beta-lactamase OS=Klebs...</td>\n",
       "      <td>tr|Q6XEC0|Q6XEC0_KLEPN</td>\n",
       "      <td>(M, R, V, L, A, L, S, A, V, F, L, V, A, S, I, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tr|Q79PC4|Q79PC4_KLEPN</td>\n",
       "      <td>tr|Q79PC4|Q79PC4_KLEPN Dihydropteroate synthas...</td>\n",
       "      <td>tr|Q79PC4|Q79PC4_KLEPN</td>\n",
       "      <td>(M, V, T, V, F, G, I, L, N, L, T, E, D, S, F, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>tr|Q7BK05|Q7BK05_KLEPN</td>\n",
       "      <td>tr|Q7BK05|Q7BK05_KLEPN EmrE OS=Klebsiella pneu...</td>\n",
       "      <td>tr|Q7BK05|Q7BK05_KLEPN</td>\n",
       "      <td>(M, K, G, W, L, F, L, V, I, A, I, V, G, E, V, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>tr|Q6SLK9|Q6SLK9_KLEPN</td>\n",
       "      <td>tr|Q6SLK9|Q6SLK9_KLEPN IS1380-like element ISE...</td>\n",
       "      <td>tr|Q6SLK9|Q6SLK9_KLEPN</td>\n",
       "      <td>(M, I, N, K, I, D, F, K, A, K, N, L, T, S, N, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>tr|Q6XEB9|Q6XEB9_KLEPN</td>\n",
       "      <td>tr|Q6XEB9|Q6XEB9_KLEPN LysR OS=Klebsiella pneu...</td>\n",
       "      <td>tr|Q6XEB9|Q6XEB9_KLEPN</td>\n",
       "      <td>(M, P, D, L, N, G, M, M, L, F, A, A, V, V, R, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>tr|Q6XEC1|Q6XEC1_KLEPN</td>\n",
       "      <td>tr|Q6XEC1|Q6XEC1_KLEPN TnpA OS=Klebsiella pneu...</td>\n",
       "      <td>tr|Q6XEC1|Q6XEC1_KLEPN</td>\n",
       "      <td>(M, R, E, A, N, I, L, Q, H, S, L, H, Q, Y, C, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>tr|Q767E2|Q767E2_KLEPN</td>\n",
       "      <td>tr|Q767E2|Q767E2_KLEPN IS6100 Transposase OS=K...</td>\n",
       "      <td>tr|Q767E2|Q767E2_KLEPN</td>\n",
       "      <td>(M, L, G, G, D, W, T, D, G, T, M, T, D, F, K, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>tr|Q8GEF1|Q8GEF1_KLEPN</td>\n",
       "      <td>tr|Q8GEF1|Q8GEF1_KLEPN Acetyltransferase (GNAT...</td>\n",
       "      <td>tr|Q8GEF1|Q8GEF1_KLEPN</td>\n",
       "      <td>(M, D, S, E, E, P, P, N, V, R, V, A, C, S, G, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>79 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                id  \\\n",
       "0   tr|A0A0A8IKC4|A0A0A8IKC4_KLEPN   \n",
       "1            sp|C7C422|BLAN1_KLEPN   \n",
       "2           tr|Q6XEC0|Q6XEC0_KLEPN   \n",
       "3           tr|Q79PC4|Q79PC4_KLEPN   \n",
       "4           tr|Q7BK05|Q7BK05_KLEPN   \n",
       "..                             ...   \n",
       "74          tr|Q6SLK9|Q6SLK9_KLEPN   \n",
       "75          tr|Q6XEB9|Q6XEB9_KLEPN   \n",
       "76          tr|Q6XEC1|Q6XEC1_KLEPN   \n",
       "77          tr|Q767E2|Q767E2_KLEPN   \n",
       "78          tr|Q8GEF1|Q8GEF1_KLEPN   \n",
       "\n",
       "                                          description  \\\n",
       "0   tr|A0A0A8IKC4|A0A0A8IKC4_KLEPN Bifunctional NA...   \n",
       "1   sp|C7C422|BLAN1_KLEPN Metallo-beta-lactamase t...   \n",
       "2   tr|Q6XEC0|Q6XEC0_KLEPN Beta-lactamase OS=Klebs...   \n",
       "3   tr|Q79PC4|Q79PC4_KLEPN Dihydropteroate synthas...   \n",
       "4   tr|Q7BK05|Q7BK05_KLEPN EmrE OS=Klebsiella pneu...   \n",
       "..                                                ...   \n",
       "74  tr|Q6SLK9|Q6SLK9_KLEPN IS1380-like element ISE...   \n",
       "75  tr|Q6XEB9|Q6XEB9_KLEPN LysR OS=Klebsiella pneu...   \n",
       "76  tr|Q6XEC1|Q6XEC1_KLEPN TnpA OS=Klebsiella pneu...   \n",
       "77  tr|Q767E2|Q767E2_KLEPN IS6100 Transposase OS=K...   \n",
       "78  tr|Q8GEF1|Q8GEF1_KLEPN Acetyltransferase (GNAT...   \n",
       "\n",
       "                              name  \\\n",
       "0   tr|A0A0A8IKC4|A0A0A8IKC4_KLEPN   \n",
       "1            sp|C7C422|BLAN1_KLEPN   \n",
       "2           tr|Q6XEC0|Q6XEC0_KLEPN   \n",
       "3           tr|Q79PC4|Q79PC4_KLEPN   \n",
       "4           tr|Q7BK05|Q7BK05_KLEPN   \n",
       "..                             ...   \n",
       "74          tr|Q6SLK9|Q6SLK9_KLEPN   \n",
       "75          tr|Q6XEB9|Q6XEB9_KLEPN   \n",
       "76          tr|Q6XEC1|Q6XEC1_KLEPN   \n",
       "77          tr|Q767E2|Q767E2_KLEPN   \n",
       "78          tr|Q8GEF1|Q8GEF1_KLEPN   \n",
       "\n",
       "                                                  seq  \n",
       "0   (M, M, D, H, T, M, T, K, N, P, D, S, I, P, Y, ...  \n",
       "1   (M, E, L, P, N, I, M, H, P, V, A, K, L, S, T, ...  \n",
       "2   (M, R, V, L, A, L, S, A, V, F, L, V, A, S, I, ...  \n",
       "3   (M, V, T, V, F, G, I, L, N, L, T, E, D, S, F, ...  \n",
       "4   (M, K, G, W, L, F, L, V, I, A, I, V, G, E, V, ...  \n",
       "..                                                ...  \n",
       "74  (M, I, N, K, I, D, F, K, A, K, N, L, T, S, N, ...  \n",
       "75  (M, P, D, L, N, G, M, M, L, F, A, A, V, V, R, ...  \n",
       "76  (M, R, E, A, N, I, L, Q, H, S, L, H, Q, Y, C, ...  \n",
       "77  (M, L, G, G, D, W, T, D, G, T, M, T, D, F, K, ...  \n",
       "78  (M, D, S, E, E, P, P, N, V, R, V, A, C, S, G, ...  \n",
       "\n",
       "[79 rows x 4 columns]"
      ]
     },
     "execution_count": 161,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def fasta_to_df(fasta_str:str)->pd.DataFrame:\n",
    "    \n",
    "    fasta_io = StringIO(fasta_str) \n",
    "    records = SeqIO.parse(fasta_io, \"fasta\")\n",
    "\n",
    "    output_list= []\n",
    "    for r in records:\n",
    "        fasta_dict ={\n",
    "        'id': r.id,\n",
    "        'description': r.description,\n",
    "        'name': r.name,\n",
    "        }\n",
    "        output_list.append(fasta_dict)\n",
    "\n",
    "    output_df =pd.DataFrame(output_list)\n",
    "\n",
    "\n",
    "#--\n",
    "fasta_to_df(all_fastas)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scicore",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.2"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
